js/ui/appDisplay.js

```diff
@@ -362,10 +362,20 @@ const EndlessApplicationView = new Lang.Class({
         return false;
     },
 
+    _reloadIcons: function() {
+        for (let icon of this._allIcons) {
+            if (icon.app) {
+                icon.app.reload_app_info();
+                icon.icon.reloadIcon();
+            }
+        }
+    },
+
     addIcons: function(isHidden) {
         // Don't do anything if we don't have more up-to-date information, since
         // re-adding icons unnecessarily can cause UX problems
         if (!this.iconsNeedRedraw()) {
+            this._reloadIcons();
```

Comment:
Just a drive-by heads up that always doing I/O from this code path unfortunately **is not** acceptable. The `iconsNeedRedraw()` system was added exactly to avoid that. It's true that reloading the visible icons only is a little better than what we were doing back then (IIRC calling `g_app_info_get_all()`) but this is still a hot path, so the smarter solution may be really what we have to do here.